fsm.go

```diff
@@ -249,6 +263,32 @@ func (f *FSM) Cannot(event string) bool {
 	return !f.Can(event)
 }
 
+// Metadata returns the value stored in data
+func (f *FSM) Metadata(key string) interface{} {
+	dataElement, ok := f.metadata[key]
```

Comment:
Lock `f.metadatMu` as a read lock here, with defer on the second line.

Reply:
I think the error can be removed also, just change the return to be the same as the map lookup `(interface{}, bool)`. It’s very uncommon to return an error as the non-error value.

Reply:
Ok, last thing: you can do `return t.metadata[key]` directly, no need for the variables here!

Reply:
I think` f.metadata[key]` returns only the value, not the bool

Reply:
No, that is incorrect, trust me: https://golang.org/ref/spec#Index_expressions

Reply:
You are correct, thanks!